### DISCUSSION TITLE
Peacock segmentation fault on MacOS

### DISCUSSION THREAD

**Discussion by brandon-barclay** [idaholab/moose#19438] [Q&A Tools] [answered]
Hello, 

I have been running into a segmentation fault when trying to run peacock on mac. Have not had any issues running moose apps but I have been unable to visualize results with peacock. I did try paraview but I was also receiving an error there about the format of my output files being incorrect, which I haven't had when using peacock on ubuntu. I am currently on MacOS 12.0.1. 

Thanks for your help.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Which version is your paraview?
The latest version complains about something we do with nodesets and sidesets (sharing the same name).
Downgrading paraview could help. The alternative is to find the legacy exodus reader inside your new paraview. I've been told that's not easy

Otherwise for peacock I dont know that we support Monterey yet. I ll try to find out

Guillaume

**Reply by milljm** [reply]
It's supposed to work yes... but I can reproduce this issue. More than likely an issue with one of dependency modules for VTK. I'll see if I can isolate it to PyQt, or Qt, etc...

```
❯ uname -r
21.1.0
```

Invoking with -i
```
❯ cd tests/kernels/simple_diffusion
❯ ../../../../python/peacock/peacock -i simple_diffusion.i
Found executable: /Users/milljm/projects/moose/test/moose_test-opt
zsh: bus error  ../../../../python/peacock/peacock -i simple_diffusion.i
```

Just running peacock.
```
❯ ../../../../python/peacock/peacock
Found executable: /Users/milljm/projects/moose/test/moose_test-opt
zsh: segmentation fault  ../../../../python/peacock/peacock
```

**Reply by brandon-barclay** [reply]
The issue I was having with paraview was the duplicate naming one, so thanks for letting me know. And as well the issue with peacock was producing the same output as above. Thanks for looking into it.

**Reply by milljm** [reply]
I can isolate this to being a PyQt/Qt issue. Unfortunately we have a dependency issue preventing us from upgrading to a newer version. While this might not be the 100% fix, this is our dilemma:

```bash
❯ mamba create -n testing_environment clangxx_osx-64=12 pyqt python=3.9

<trimmed>

Looking for: ['clangxx_osx-64=12', 'pyqt', 'python=3.9']

conda-forge/osx-64       Using cache
conda-forge/noarch       Using cache
public/osx-64            [====================] (00m:00s) No change
public/noarch            [====================] (00m:00s) No change
Encountered problems while solving:
  - package clangxx_osx-64-12.0.1-h7e1b574_0 requires clang_osx-64 12.0.1 hb91bd55_0, but none of the providers can be installed
```

While not very forthcoming with the actual dependency error; What the above is trying to tell us is that there exists no version of PyQt for any version of Python higher than 3.7 when also requiring Clang 12.

The gist of it is, we have moved on to Clang 12.x compilers, while PyQt has _only_ been built with Clang 12.x for Python 3.7. I think the fix exists in later versions of Python/PyQt/Qt. I will attempt to see if that is the case. Will take some doing (use older version of MOOSE, since Peacock requires pyhit, which is a moose contrib).

Indeed, throughout this whole process, I've managed to get Peacock to launch correctly 4 times (out of many attempts while using `pdb`), with no changes to any code (thread buggy?). Seems to be something with:

```diff
diff --git a/python/peacock/Execute/ExecuteOptionsPlugin.py b/python/peacock/Execute/ExecuteOptionsPlugin.py
index 480793cfca..6df59acf86 100755
--- a/python/peacock/Execute/ExecuteOptionsPlugin.py
+++ b/python/peacock/Execute/ExecuteOptionsPlugin.py
@@ -168,12 +168,12 @@ class ExecuteOptionsPlugin(QWidget, Plugin):
         self._loading_dialog.setInformativeText(app_path)
         self._loading_dialog.show()
         self._loading_dialog.raise_()
-        QApplication.processEvents()
+        QApplication.sendPostedEvents()

         app_info = ExecutableInfo()
```
That is, something is dying in a PyQt thread process when asking it to perform background tasks (ref: https://doc.qt.io/qt-5/qcoreapplication.html#processEvents).

**Reply by milljm** [reply]
Confirmed... while pinning a later Python (3.9), Peacock seems to launch. So far, this is what I am having to do:

```bash
❯ mamba init             # I am using Mambaforge, replace your commands with `conda` where applicable
< restart terminal >
❯ mamba create -n peacock python=3.9 clangxx_osx-64 clang_osx-64 \
  gfortran_osx-64 mpich boa conda-build pyqt pandas matplotlib vtk
❯ mamba activate peacock
< make changes to moose/conda/pyhit, see below diff >
❯ cd moose/conda/pyhit
❯ mamba mambabuild .     # or `conda build .`
❯ mamba install --use-local moose-pyhit
❯ mamba deactivate
❯ mamba activate peacock # reload environment
❯ cd moose/python
❯ mv pyhit not_pyhit     # Looks like Peacock prepends this path to syspath, making it impossible
                         # to load our custom moose-pyhit package instead
❯ cd peacock
❯ ./peacock              # profit. So far so good. Peacock loads at least.
```

pyhit package diff (force build pyhit with Clang 11.x and Python 3.9):
```diff
diff --git a/conda/pyhit/meta.yaml b/conda/pyhit/meta.yaml
index 854b57f881..6f252f7792 100644
--- a/conda/pyhit/meta.yaml
+++ b/conda/pyhit/meta.yaml
@@ -1,5 +1,5 @@
 {% set build = 0 %}
-{% set version = "2021.10.27" %}
+{% set version = "2021.11.29" %}
 
 package:
   name: moose-pyhit
@@ -14,18 +14,18 @@ build:
 
 requirements:
   build:
-    - {{ moose_cxx }}
-    - {{ moose_python }}
+    - clangxx_osx-64 11.1.0
+    - python 3.9
     - cython
 
   host:
-    - {{ moose_cxx }}
-    - {{ moose_python }}
+    - clangxx_osx-64 11.1.0
+    - python 3.9
     - cython
 
   run:
-    - {{ moose_cxx }}
-    - python
+    - clangxx_osx-64 11.1.0
+    - python 3.9
 
 test:
   imports:
```

With the above partially working, I think it might be possible to use a different environment solely for the purpose of launching Peacock. I'll continue to investigate. You may be thinking; Why not just use Clang 11? Well, Monterey's XCode version(s) do not gibe well with Clang-11. It forced us to move to Clang 12... MOOSE simply doesn't compile correctly under Monterey with Clang-11. @cticenhour is that correct?

@brandon-barclay I am not exactly asking you to perform the above as a fix. I am just documenting my findings, as I find them. Although if you do wish to build the above, you will need to obtain the 10.9 MacSDK and install it to /opt. I use the conveniently available SDKs from here: https://github.com/phracker/MacOSX-SDKs/releases

**Reply by cticenhour** [reply]
Monterey (or Big Sur with the most up-to-date Xcode version) just did not work with Clang 11, presumably due to the changes needed in order for Apple to have compatibility with their newer Apple Silicon platforms. The end result was that much of the MOOSE dependencies could be built successfully, but the final build of MOOSE would be riddled with library and linking errors resulting in a broken build. It's very possible Clang-11 could have worked with some heavy effort on Mac Intel systems, but our push for Apple Silicon support preempted any more work on it. Clang 12.0.0 or later holds partially optimized support for that architecture, with more full-featured support coming in Clang-13.

**Reply by milljm** [reply]
Looks like using two environments may be the only path forward for now. Also, renaming that "moose/python/pyhit" directory is necessary. I'll have to bring all this up with my group, on how to proceed. I like the idea of no longer tying Peacock to moose/python/pyhit. As it seems to work just fine on my end:

![Screen Shot 2021-11-29 at 11 02 09 AM](https://user-images.githubusercontent.com/6414286/143919863-4686c29c-54e6-4847-bd63-ef74983f8c6b.png)

![Screen Shot 2021-11-29 at 11 02 20 AM](https://user-images.githubusercontent.com/6414286/143919880-625154d0-e991-4fd6-bf1a-b8bfb81a09fa.png)

![Screen Shot 2021-11-29 at 11 02 31 AM](https://user-images.githubusercontent.com/6414286/143919909-73fab4cc-20ed-4d04-ad6c-f70ea078ce8e.png)
